lang/shared/src/main/scala/com/wavesplatform/lang/Evaluator.scala

```diff
@@ -1,169 +1,140 @@
 package com.wavesplatform.lang
 
+import cats.data.EitherT
+import com.wavesplatform.lang.Context.Obj
 import com.wavesplatform.lang.Terms._
-import scodec.bits.ByteVector
-import scorex.crypto.signatures.{Curve25519, PublicKey, Signature}
+import monix.eval.Coeval
 
 import scala.util.{Failure, Success, Try}
 
 object Evaluator {
 
-  import scala.util.control.TailCalls.{TailRec, done, tailcall}
-
-  type Defs = Map[String, (TYPE, Any)]
-
-  case class Context(typeDefs: Map[String, CUSTOMTYPE], defs: Defs)
-  object Context {
-    val empty = Context(Map.empty, Map.empty)
-  }
-
-  type TypeResolutionError = String
-  type ExcecutionError = String
-  type ExecResult[T] = Either[ExcecutionError, T]
-
-  private def r[T](ctx: Context, t: Typed.EXPR): TailRec[ExecResult[T]] = {
-    (t match {
-      case Typed.BLOCK(mayBeLet, inner, blockTpe) => tailcall {
-        mayBeLet match {
-          case None =>
-            r[blockTpe.Underlying](ctx, inner)
-
-          case Some(Typed.LET(newVarName, newVarBlock)) =>
-            ctx.defs.get(newVarName) match {
-              case Some(_) => done(Left(s"Value '$newVarName' already defined in the scope"))
-              case None =>
-                val varBlockTpe = newVarBlock.tpe
-                  r[varBlockTpe.Underlying](ctx, newVarBlock).flatMap {
-                    case Left(e) => done(Left(e))
-                    case Right(newVarValue) =>
-                      val updatedCtx = ctx.copy(defs = ctx.defs.updated(newVarName, (varBlockTpe, newVarValue)))
-                      r[blockTpe.Underlying](updatedCtx, inner)
+  type TypeResolutionError      = String
+  type ExcecutionError          = String
+  type ExecResult[T]            = Either[ExcecutionError, T]
+  type TrampolinedExecResult[T] = EitherT[Coeval, ExcecutionError, T]
+
+  private def r[T](ctx: Context, t: TrampolinedExecResult[Typed.EXPR]): TrampolinedExecResult[T] =
+    t flatMap { (typedExpr: Typed.EXPR) =>
+      (typedExpr match {
+        case Typed.BLOCK(mayBeLet, inner, blockTpe) =>
+          mayBeLet match {
+            case None => r[blockTpe.Underlying](ctx, EitherT.pure(inner))
+            case Some(Typed.LET(newVarName, newVarBlock)) =>
+              ctx.letDefs.get(newVarName) match {
+                case Some(_) => EitherT.leftT[Coeval, T](s"Value '$newVarName' already defined in the scope")
+                case None =>
+                  val varBlockTpe                                                  = newVarBlock.tpe
+                  val eitherTCoeval: TrampolinedExecResult[varBlockTpe.Underlying] = r[varBlockTpe.Underlying](ctx, EitherT.pure(newVarBlock))
+                  val calculatedValue: Coeval[varBlockTpe.Underlying] = eitherTCoeval.value.flatMap {
+                    case Left(err) => Coeval(???)
```

Comment:
So, in a runtime we will see "Not implement error". I think, it rather better to save the error for the later usage: when the expression is evaluating, and Coeval contains an error, the whole expression should fail with this error.

Please create a task in Jira, if you agree with me :)

Reply:
yes, forgot about this `???`. you're right.